core/templates/dev/head/pages/exploration_player/AudioControlsDirective.js

```diff
@@ -174,15 +162,43 @@ oppia.directive('audioControls', [
               AudioPlayerService.getCurrentAudioControlsDirectiveId();
           };
 
+          var onFinishedLoadingAudio = function(loadedFilename) {
+            var audioTranslation = getAudioTranslationInCurrentLanguage();
+            if (requestedToPlayAudio && loadedFilename ===
+                audioTranslation.filename) {
+              requestedToPlayAudio = false;
+              AudioPlayerService.load(audioTranslation.filename, directiveId)
+                .then(function() {
+                  AudioPlayerService.play();
+                });
+            }
+            $scope.audioIsLoading = false;
+          };
+
+          if (!AudioPreloaderService.isRunning()) {
```

Comment:
This pattern also bothers me a little. Originally, I was calling kickOffAudioPreloader() inside ExplorationPlayerService init, but then I have no way to pass in the necessary callback function.

Reply:
I would expect the kickoff to be in ExplorationPlayerService.init, yeah.

I'm a bit confused about why you need the callback function. My mental model is that the thing that requests audio is the one that needs a callback function (i.e. when the audio button is pressed). But I don't see why the kickoff triggerer needs it. (Basically, I'm thinking there's a kickoff triggered when the exploration is loaded, and subsequent individual requests that try to jump the queue.)

Reply:
So this main edge case being handled here is this:

1.) Audio preloading kicks off at start of exploration
2.) Learners presses speaker icon before audio is loaded

If I start the kickoff in ExplorationPlayerService.init I'm having trouble seeing how I can start playing the audio once the audio has loaded (other than via a broadcast).

Reply:
If we go with that solution, I suppose we might want to also be able to nullify all other callbacks if a different audio button is subsequently pressed.

Another approach is, when an audio file finishes downloading, have a callback that checks the filename of the "latest audio request" and if the filename matches that of the file that's just finished downloading, execute playback.